**User based Collaborative Filtering**

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
ratings=pd.read_csv('/content/drive/My Drive/Dataset/Ratings .csv')
movies=pd.read_csv('/content/drive/My Drive/Dataset/movies.csv')
userdata=pd.read_csv('/content/drive/My Drive/Dataset/user_data.csv')

In [ ]:
movies=pd.read_csv('/content/drive/My Drive/Dataset/movies.csv')
movies

MovieID                               Title                        Genres
0           1                    Toy Story (1995)   Animation|Children's|Comedy
1           2                      Jumanji (1995)  Adventure|Children's|Fantasy
2           3             Grumpier Old Men (1995)                Comedy|Romance
3           4            Waiting to Exhale (1995)                  Comedy|Drama
4           5  Father of the Bride Part II (1995)                        Comedy
...       ...                                 ...                           ...
3878     3948             Meet the Parents (2000)                        Comedy
3879     3949          Requiem for a Dream (2000)                         Drama
3880     3950                    Tigerland (2000)                         Drama
3881     3951             Two Family House (2000)                         Drama
3882     3952               Contender, The (2000)                Drama|Thriller

[3883 rows x 3 columns]

In [ ]:
ratings=pd.read_csv('/content/drive/My Drive/Dataset/Ratings .csv')
ratings

UserID  MovieID  Rating  Timestamp
0             1     1193       5  978300760
1             1      661       3  978302109
2             1      914       3  978301968
3             1     3408       4  978300275
4             1     2355       5  978824291
...         ...      ...     ...        ...
1000204    6040     1091       1  956716541
1000205    6040     1094       5  956704887
1000206    6040      562       5  956704746
1000207    6040     1096       4  956715648
1000208    6040     1097       4  956715569

[1000209 rows x 4 columns]

In [ ]:
userdata=pd.read_csv('/content/drive/My Drive/Dataset/user_data.csv')
userdata

UserID Gender  Age  Occupation Zip-code
0          1      F    1          10    48067
1          2      M   56          16    70072
2          3      M   25          15    55117
3          4      M   45           7     2460
4          5      M   25          20    55455
...      ...    ...  ...         ...      ...
6035    6036      F   25          15    32603
6036    6037      F   45           1    76006
6037    6038      F   56           1    14706
6038    6039      F   45           0     1060
6039    6040      M   25           6    11106

[6040 rows x 5 columns]

In [ ]:
df=ratings.merge(userdata, on='UserID',how='left').merge(movies, on='MovieID',how='left')
df

UserID  ...                           Genres
0             1  ...                            Drama
1             1  ...     Animation|Children's|Musical
2             1  ...                  Musical|Romance
3             1  ...                            Drama
4             1  ...      Animation|Children's|Comedy
...         ...  ...                              ...
1000204    6040  ...                           Comedy
1000205    6040  ...                Drama|Romance|War
1000206    6040  ...                     Comedy|Drama
1000207    6040  ...                            Drama
1000208    6040  ...  Children's|Drama|Fantasy|Sci-Fi

[1000209 rows x 10 columns]

In [ ]:
df=df.drop(columns='Timestamp')
df

UserID  ...                           Genres
0             1  ...                            Drama
1             1  ...     Animation|Children's|Musical
2             1  ...                  Musical|Romance
3             1  ...                            Drama
4             1  ...      Animation|Children's|Comedy
...         ...  ...                              ...
1000204    6040  ...                           Comedy
1000205    6040  ...                Drama|Romance|War
1000206    6040  ...                     Comedy|Drama
1000207    6040  ...                            Drama
1000208    6040  ...  Children's|Drama|Fantasy|Sci-Fi

[1000209 rows x 9 columns]

**SVD (Singular Value Decomposition )**

In [ ]:
final_df = ratings.pivot(index='UserID',columns='MovieID',values='Rating')

In [ ]:
final_df.fillna(0,inplace=True)
final_df

MovieID  1     2     3     4     5     6     ...  3947  3948  3949  3950  3951  3952
UserID                                       ...                                    
1         5.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   0.0   0.0   0.0   0.0   0.0
2         0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   0.0   0.0   0.0   0.0   0.0
3         0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   0.0   0.0   0.0   0.0   0.0
4         0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   0.0   0.0   0.0   0.0   0.0
5         0.0   0.0   0.0   0.0   0.0   2.0  ...   0.0   0.0   0.0   0.0   0.0   0.0
...       ...   ...   ...   ...   ...   ...  ...   ...   ...   ...   ...   ...   ...
6036      0.0   0.0   0.0   2.0   0.0   3.0  ...   0.0   0.0   0.0   0.0   0.0   0.0
6037      0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   0.0   0.0   0.0   0.0   0.0
6038      0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   0.0   0.0   0.0   0.0   0.0
6039      0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   0.0   0.0   0.0   0.0   0.0
6040      3.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   0.0   0.0   0.0   0.0   0.0

[6040 rows x 3706 columns]

In [ ]:
rel = final_df.values
meann = np.mean(rel, axis = 1)
final_mean = rel - meann.reshape(-1, 1)

In [ ]:
from scipy.sparse.linalg import svds
U, sig, Vt = svds(final_mean, k = 50)

In [ ]:
sig = np.diag(sig)

In [ ]:
all_user = np.dot(np.dot(U, sig), Vt) + meann.reshape(-1, 1)

In [ ]:
pred_df = pd.DataFrame(all_user, columns = final_df.columns)
pred_df.head()


MovieID      1         2         3     ...      3950      3951      3952
0        4.288861  0.143055 -0.195080  ...  0.031912  0.050450  0.088910
1        0.744716  0.169659  0.335418  ... -0.101102 -0.054098 -0.140188
2        1.818824  0.456136  0.090978  ...  0.012345  0.015148 -0.109956
3        0.408057 -0.072960  0.039642  ... -0.026050  0.014841 -0.034224
4        1.574272  0.021239 -0.051300  ...  0.033830  0.125706  0.199244

[5 rows x 3706 columns]

In [ ]:
pred_df.shape

(6040, 3706)

In [ ]:
def recommend_movies(userID, n=10):
    
    # Get and sort the user's predictions
    user_idx = userID - 1 # UserID starts at 1, not 0
    user_pred = pred_df.iloc[user_idx].sort_values(ascending=False) # UserID starts at 1
    
    # Get the user's data and merge in the movie information.
    user_data = ratings[ratings.UserID == (userID)]
    user_full = (user_data.merge(movies, how = 'left', left_on = 'MovieID', right_on = 'MovieID').sort_values(['Rating'], ascending=False))
    # Recommend the highest predicted rating movies that the user hasn't seen yet.
    recommendations = (movies[~movies['MovieID'].isin(user_full['MovieID'])].merge(pd.DataFrame(user_pred).reset_index(), how = 'left', left_on = 'MovieID', right_on = 'MovieID').rename(columns = {user_idx: 'Predictions'}).sort_values('Predictions', ascending = False).iloc[:n, :-1])
    
    print ("User {0} has already rated {1} movies.".format(userID, user_full.shape[0]))
    print ("Recommending highest {0} predicted rating movies not already rated.".format(n))
    pd.set_option('display.max_columns', None)
    return(recommendations)

In [ ]:
recommend_movies(1,10)

User 1 has already rated 53 movies.
Recommending highest 10 predicted rating movies not already rated.
      MovieID                             Title  \
311       318  Shawshank Redemption, The (1994)   
32         34                       Babe (1995)   
356       364             Lion King, The (1994)   
1975     2081        Little Mermaid, The (1989)   
1235     1282                   Fantasia (1940)   
1974     2080         Lady and the Tramp (1955)   
1972     2078           Jungle Book, The (1967)   
1990     2096            Sleeping Beauty (1959)   
1981     2087                  Peter Pan (1953)   
348       356               Forrest Gump (1994)   

                                           Genres  
311                                         Drama  
32                        Children's|Comedy|Drama  
356                  Animation|Children's|Musical  
1975  Animation|Children's|Comedy|Musical|Romance  
1235                 Animation|Children's|Musical  
1974  Animation|Childre

In [ ]:
dd=recommend_movies(1,10)

User 1 has already rated 53 movies.
Recommending highest 10 predicted rating movies not already rated.


In [ ]:
dd

MovieID                             Title  \
311       318  Shawshank Redemption, The (1994)   
32         34                       Babe (1995)   
356       364             Lion King, The (1994)   
1975     2081        Little Mermaid, The (1989)   
1235     1282                   Fantasia (1940)   
1974     2080         Lady and the Tramp (1955)   
1972     2078           Jungle Book, The (1967)   
1990     2096            Sleeping Beauty (1959)   
1981     2087                  Peter Pan (1953)   
348       356               Forrest Gump (1994)   

                                           Genres  
311                                         Drama  
32                        Children's|Comedy|Drama  
356                  Animation|Children's|Musical  
1975  Animation|Children's|Comedy|Musical|Romance  
1235                 Animation|Children's|Musical  
1974  Animation|Children's|Comedy|Musical|Romance  
1972          Animation|Children's|Comedy|Musical  
1990                 Animation|Children's|Musical  
1981         Animation|Children's|Fantasy|Musical  
348                            Comedy|Romance|War

SVD ( using surprise)

In [ ]:
!pip install surprise

     |████████████████████████████████| 6.5MB 3.0MB/s 
  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.0-cp36-cp36m-linux_x86_64.whl size=1675740 sha256=95e4052b285e258fe4b4f0c3a98d8279ee6be9e8611d3763e4671de40bac037f
  Stored in directory: /root/.cache/pip/wheels/cc/fa/8c/16c93fccce688ae1bde7d979ff102f7bee980d9cfeb8641bcf
Successfully built scikit-surprise


In [ ]:
from collections import defaultdict

from surprise import SVD
from surprise import Dataset

In [ ]:
def get_top_n(predictions, n=10):
# First map the predictions to each user.
    top_n = defaultdict(list)
    for UserID, MovieID, true_r, est, _ in predictions:
        top_n[UserID].append((MovieID, est))
        # Then sort the predictions for each user and retrieve the k highest ones.
    for UserID, ratings in top_n.items():
        ratings.sort(key=lambda x: x[1], reverse=True)
        top_n[UserID] = ratings[:n]
    return top_n

In [ ]:
# First train an SVD algorithm on the movielens dataset.
from surprise import Reader
reader = Reader()
data = Dataset.load_from_df(df[['UserID', 'MovieID', 'Rating']], reader)
trainset = data.build_full_trainset()
algo = SVD()
algo.fit(trainset)

In [ ]:
# Than predict ratings for all pairs (u, i) that are NOT in the training set.
testset = trainset.build_anti_testset()
predictions = algo.test(testset)

In [ ]:
top_n = get_top_n(predictions, n=10)

In [ ]:
# Print the recommended items for each user
for UserID, ratings in top_n.items():
    print(UserID, [MovieID for (MovieID, _) in ratings])

Streaming output truncated to the last 5000 lines.
1042 [1610, 318, 1262, 527, 1207, 2905, 110, 670, 912, 2762]
1043 [919, 910, 2905, 1028, 745, 1207, 1254, 926, 969, 920]
1044 [750, 2360, 3134, 2351, 1178, 922, 2019, 50, 3435, 1212]
1045 [356, 318, 527, 914, 953, 2324, 2905, 1035, 1242, 2268]
1046 [2905, 745, 2609, 1172, 50, 318, 3470, 541, 260, 1200]
1047 [527, 318, 3338, 2905, 1949, 2324, 50, 745, 1193, 1148]
1048 [2905, 541, 858, 3435, 1204, 1250, 1262, 1198, 2858, 1267]
1049 [1256, 3307, 1207, 1278, 1148, 3030, 2905, 318, 1193, 1223]
1050 [1394, 2905, 2019, 1036, 3114, 3338, 858, 1136, 1307, 2918]
1051 [1296, 1023, 3307, 1131, 2905, 955, 1214, 1069, 3350, 1283]
1052 [2905, 745, 2019, 670, 3022, 260, 904, 905, 3338, 720]
1053 [912, 1262, 923, 593, 318, 953, 1278, 2728, 1213, 908]
1054 [527, 2762, 150, 260, 1207, 2028, 1246, 1210, 3147, 110]
1055 [912, 923, 2019, 750, 919, 1207, 1358, 2905, 608, 858]
1056 [3671, 1947, 2324, 932, 2301, 3462, 306, 144, 527, 1949]
1057 [1204, 912, 898,

In [ ]:
#6040 [1204, 3435, 3307, 1178, 922, 307, 2186, 1260, 1267, 926]
movies.loc[movies['MovieID'].isin(['1204','3435','3307','1178','922','307','2186','1260','1267','926'])]

MovieID  ...                    Genres
304       307  ...                     Drama
910       922  ...                 Film-Noir
914       926  ...                     Drama
1162     1178  ...                 Drama|War
1186     1204  ...             Adventure|War
1240     1260  ...  Crime|Film-Noir|Thriller
1247     1267  ...        Film-Noir|Thriller
2117     2186  ...        Film-Noir|Thriller
3238     3307  ...      Comedy|Drama|Romance
3366     3435  ...           Crime|Film-Noir

[10 rows x 3 columns]

In [ ]:
from surprise import accuracy
accuracy.rmse(predictions, verbose=True)

RMSE: 0.6972


0.6971733375216481

In [ ]:
accuracy.mae(predictions, verbose=True)

MAE:  0.5402


0.540232292839506

 **NMF ( Non-Negative Matrix Factorization )**

In [ ]:
from surprise import NMF
from surprise import Dataset
from collections import defaultdict

In [ ]:
def get_top_n(predictions, n=10):
# First map the predictions to each user.
    top_n = defaultdict(list)
    for uid, mid, true_r, est, _ in predictions:
        top_n[uid].append((mid, est))
        # Then sort the predictions for each user and retrieve the k highest ones.
    for uid, ratings in top_n.items():
        ratings.sort(key=lambda x: x[1], reverse=True)
        top_n[uid] = ratings[:n]
    return top_n

In [ ]:
# First train an SVD algorithm on the movielens dataset.
from surprise import Reader
reader = Reader()
data = Dataset.load_from_df(df[['UserID', 'MovieID', 'Rating']], reader)
trainset = data.build_full_trainset()
algo = NMF()
algo.fit(trainset)

In [ ]:
# Than predict ratings for all pairs (u, i) that are NOT in the training set.
testset = trainset.build_anti_testset()
predictions = algo.test(testset)

In [ ]:
top_n = get_top_n(predictions, n=10)

In [ ]:
# Print the recommended items for each user
for uid, ratings in top_n.items():
    print(uid, [mid for (mid, _) in ratings])

Streaming output truncated to the last 5000 lines.
1042 [557, 3306, 1743, 572, 53, 787, 3233, 3092, 1207, 922]
1043 [787, 2503, 557, 572, 3306, 3092, 3880, 578, 922, 1204]
1044 [578, 557, 572, 3382, 2503, 787, 53, 3245, 3233, 2930]
1045 [858, 3338, 787, 2905, 1471, 1664, 578, 2342, 1117, 2981]
1046 [3092, 787, 854, 3245, 2999, 3517, 557, 572, 1423, 1002]
1047 [1207, 1250, 2904, 649, 2905, 2999, 1423, 1164, 578, 3410]
1048 [578, 557, 572, 53, 3233, 1471, 853, 1131, 2503, 3306]
1049 [557, 572, 649, 1851, 3338, 2503, 3245, 3172, 1138, 3382]
1050 [557, 53, 578, 2503, 2198, 787, 572, 3233, 2309, 649]
1051 [578, 2503, 1420, 557, 3233, 572, 3382, 53, 2905, 2197]
1052 [3245, 578, 2197, 2503, 1534, 557, 3233, 572, 1423, 858]
1053 [557, 578, 572, 3233, 2503, 1237, 787, 3245, 527, 2562]
1054 [1193, 914, 1197, 1287, 2804, 919, 2398, 2918, 720, 1270]
1055 [557, 572, 578, 1423, 2503, 2197, 3517, 1131, 3233, 3382]
1056 [3092, 2904, 214, 669, 854, 2351, 3245, 2999, 1471, 1423]
1057 [557, 572, 2905, 23

In [ ]:
a=int(input("Enter UserID : "))
top_n[a]

Enter UserID : 6040


[(557, 5),
 (53, 4.611415143722972),
 (3245, 4.580747866608285),
 (578, 4.5755220442708815),
 (1149, 4.546701085174968),
 (3172, 4.453425795466548),
 (1178, 4.440149905346194),
 (2931, 4.4147885685206605),
 (572, 4.393983580169678),
 (3022, 4.374197022135702)]

In [ ]:
movies.loc[movies['MovieID'].isin(['557','53','3245','578','1149','3172','1178','2931','572','3022'])]

MovieID                                        Title             Genres
52         53                              Lamerica (1994)              Drama
553       557                            Mamma Roma (1962)              Drama
568       572                       Foreign Student (1994)              Drama
574       578                  Hour of the Pig, The (1993)      Drama|Mystery
1133     1149    JLG/JLG - autoportrait de d�cembre (1994)  Documentary|Drama
1162     1178                        Paths of Glory (1957)          Drama|War
2862     2931  Time of the Gypsies (Dom za vesanje) (1989)              Drama
2953     3022                          General, The (1927)             Comedy
3103     3172                      Ulysses (Ulisse) (1954)          Adventure
3176     3245          I Am Cuba (Soy Cuba/Ya Kuba) (1964)              Drama

In [ ]:
# for userID 1
a=int(input("Enter UserID : "))
top_n[a]

Enter UserID : 1


[(3089, 5),
 (649, 5),
 (3245, 5),
 (1743, 5),
 (578, 5),
 (2931, 5),
 (2503, 5),
 (53, 5),
 (3517, 5),
 (557, 5)]

In [ ]:
movies.loc[movies['MovieID'].isin(['3089','649','3245','1743','578','2931','2503','53','3517','557'])]

MovieID                                            Title         Genres
52         53                                  Lamerica (1994)          Drama
553       557                                Mamma Roma (1962)          Drama
574       578                      Hour of the Pig, The (1993)  Drama|Mystery
643       649               Cold Fever (� k�ldum klaka) (1994)   Comedy|Drama
1692     1743                         Arguing the World (1996)    Documentary
2434     2503                          Apple, The (Sib) (1998)          Drama
2862     2931      Time of the Gypsies (Dom za vesanje) (1989)          Drama
3020     3089  Bicycle Thief, The (Ladri di biciclette) (1948)          Drama
3176     3245              I Am Cuba (Soy Cuba/Ya Kuba) (1964)          Drama
3448     3517                                Bells, The (1926)    Crime|Drama

In [ ]:
from surprise import accuracy
accuracy.rmse(predictions, verbose=True)

RMSE: 0.8199


0.8198827049186845

In [ ]:
accuracy.mae(predictions, verbose=True)

MAE:  0.6426


0.6425886682286809

 Here, SVD with RMSE: 0.6972 , NMF with RMSE: 0.8199

 SO, SVD gives better performance/results than NMF.

***Maths behind the matrix factorisation***

Matrix factorization is a class of collaborative filtering algorithms used in recommender systems. Matrix factorization algorithms work by decomposing the user-item interaction matrix into the product of two lower dimensionality rectangular matrices.

The idea behind matrix factorization is to represent users and items in a lower dimensional latent space.

Matrix factorization can be seen as breaking down a large matrix into a product of smaller ones.


        Rating Matrix[R] = User Matrix[P] * Item Matrix[Q](transpose)

 R(m user * n Items) = P(m user * k features) * Q(k feature * n items)(trans.)

 ***Singular value decomposition ( SVD )***

 It decomposes 
R
 into two unitary matrices and a diagonal matrix:

R = U Σ V (Transpose)

where 
R
 is user ratings matrix, 
U
 is the user “features” matrix, 
Σ
 is the diagonal matrix of singular values (essentially weights), and 
V
( Transpose)
 is the movie “features” matrix. 
U
 and 
V
( Transpose )
 are orthogonal, and represent different things. 
U
 represents how much users “like” each feature and 
V
( Transpose )
 represents how relevant each feature is to each movie.

  ***Non-Negative Matrix Factorization ( NMF )***

  NMF decomposes the non-negative utility matrix R into the product of matrices
   W and H: 
R (n ∗ d) = W (n ∗ r) . H (r ∗ d)

Where columns in matrix 
W (n ∗ r)
 represent components, while matrix 
H (r ∗ d)
 stores the corresponding weights. More importantly, NMF introduces constraints under which: 
W
≥
0
 and 
H
≥
0
.The component-wise nonnegativity is a substantial difference from SVD. 
